modules/reindex/src/main/java/org/elasticsearch/index/reindex/remote/RemoteScrollableHitSource.java

```diff
@@ -0,0 +1,231 @@
+        class RetryHelper extends AbstractRunnable {
+            private final Iterator<TimeValue> retries = backoffPolicy.iterator();
+
+            @Override
+            protected void doRun() throws Exception {
+                client.performRequest(method, uri, params, entity, new ResponseListener() {
+                    @Override
+                    public void onResponse(InputStream content) {
+                        T response;
+                        try {
+                            XContent xContent = XContentFactory.xContentType(content).xContent();
+                            try(XContentParser xContentParser = xContent.createParser(content)) {
+                                response = parser.apply(xContentParser, () -> ParseFieldMatcher.STRICT);
+                            }
+                        } catch (IOException e) {
+                            throw new ElasticsearchException("Error deserializing response", e);
+                        }
+                        listener.accept(response);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        fail.accept(e);
+                    }
+
+                    @Override
+                    public void onRetryableFailure(Throwable t) {
+                        if (retries.hasNext()) {
+                                TimeValue delay = retries.next();
+                                logger.trace("retrying rejected search after [{}]", t, delay);
+                                countSearchRetry.run();
+                                threadPool.schedule(delay, ThreadPool.Names.SAME, RetryHelper.this);
+                        } else {
+                            fail.accept(t);
+                        }
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                fail.accept(t);
+            }
+        }
+        new RetryHelper().run();
+    }
+
+    public interface AsyncClient extends Closeable {
+        void performRequest(String method, String uri, Map<String, String> params, HttpEntity entity, ResponseListener listener);
+    }
+
+    public interface ResponseListener extends ActionListener<InputStream> {
+        void onRetryableFailure(Throwable t);
+    }
+
+    public static class AsynchronizingRestClient implements AsyncClient {
+        private final ThreadPool threadPool;
+        private final RestClient restClient;
+
+        public AsynchronizingRestClient(ThreadPool threadPool, RestClient restClient, Consumer<Throwable> fail) {
+            this.threadPool = threadPool;
+            this.restClient = restClient;
+        }
+
+        @Override
+        public void performRequest(String method, String uri, Map<String, String> params, HttpEntity entity,
+                ResponseListener listener) {
+            threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
It is the only thread pool we can block a thread in. I'll leave a comment. I expect we'll get an async client before too long and it'll be moot but I still think leaving a note is a good idea.